cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AccessEnforcerV2.java

```diff
@@ -0,0 +1,102 @@
+   * on the specified {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permission
+   * @param permission the {@link Permission} being performed
+   *
+   * @return EnforcementResult
+   */
+  default EnforcementResult enforce(EntityId entity, Principal principal, Permission permission)
+    throws AccessException {
+    return enforce(entity, principal, Collections.singleton(permission));
+  }
+
+  /**
+   * Enforces authorization for the specified {@link Principal} for the specified {@link Permission
+   * permissions} on the specified {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permissions
+   * @param permissions the {@link Permission permissions} being performed
+   * Should NOT throw  UnauthorizedException if the principal is not authorized to perform the specified
+   *     permissions on the entity
+   *
+   * @return EnforcementResult
+   */
+  EnforcementResult enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException;
+
+  /**
+   * Enforces specific {@link Permission#isCheckedOnParent()} permission for {@link EntityType} on
+   * it's parent {@link EntityId}. E.g. one can check if it's possible to {@link
+   * StandardPermission#LIST} {@link EntityType#PROFILE} in specific {@link
+   * io.cdap.cdap.proto.id.NamespaceId}.
+   *
+   * @param entityType the {@link EntityType} on which authorization is to be enforced
+   * @param parentId the {@link EntityId} of parent entity on which authorization is to be
+   *     enforced
+   * @param principal the {@link Principal} that performs the permission
+   * @param permission the {@link Permission} being performed. Permission must return true on
+   *     {@link Permission#isCheckedOnParent()}.
+   * Should NOT throw UnauthorizedException if the principal is not authorized to perform the specified
+   *     permission on the entity
+   *
+   * @return EnforcementResult
+   */
+  EnforcementResult enforceOnParent(EntityType entityType, EntityId parentId, Principal principal,
+                       Permission permission) throws AccessException;
+
+  /**
+   * Checks whether the set of {@link EntityId}s are visible to the specified {@link Principal}. An
+   * entity is visible to a principal if the principal has {@link StandardPermission#GET} on the
+   * entity. However, visibility check behavior can be overwritten at the authorization extension
+   * level.
+   *
+   * @param entityIds the entities on which the visibility check is to be performed
+   * @param principal the principal to check the visibility for
+   * @return a set of entities that are visible to the principal
+   */
+  Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
```

Comment:
I am wondering how do you plan to handle this? You would need at least one EnforcementResult if not a result for each entity.

Reply:
I think we can have a `Map<? extends EntityId, AuthorizationResult>` and then in `DefaultAccessEnforcer`'s isVisible, we can filter based on each entity’s `AuthorizationResult.AuthorizationStatus`  != `AuthorizationStatus.NO`

Reply:
Yes. Something along the lines. And we should do CAL if any of the enities has it enabled